text/0000-derive-bound-control.md

```diff
@@ -0,0 +1,829 @@
+
+```rust
+macro_rules! Foo { () => { T } }
+struct Bar<T>(Foo!())
+```
+
+macros have no way to expand `Foo!()`. Arguably, using type position macros
+are rare, but for standardization, a more robust approach is probably preferred.
+A possibly path ahead is to provide the API proposed  in [RFC 2320], in which
+case using the field based approach becomes more robust.
+
+[RFC 2320]: https://github.com/rust-lang/rfcs/pull/2320
+
+There's also the question of whether interpretations of `#[derive_no_bound]`
+on fields is legible and intuitive, which misunderstandings so far during
+development of this RFC has shown is not unlikely.
+
+## 2. Should `#[derive_field_bound]` and `#[derive_no_bound]` be combinable?
+
+Consider the following snippet:
+
+```rust
+#[derive(Clone, PartialEq, PartialOrd)]
+struct Foo<T> {
+    #[derive_field_bound]
+    #[derive_no_bound(Clone)]
+    field: Bar<T>
+}
+```
+
+This could be interpreted as an instruction to provide the following `impl`s:
+
+```rust
+impl<T> Clone for Foo<T> {..}
+impl<T: PartialEq> PartialEq for Foo<T> where Bar<T>: PartialEq {..}
+impl<T: PartialOrd> PartialOrd for Foo<T> where Bar<T>: PartialOrd {..}
+```
+
+This is currently not proposed as it is deemed unnecessary, but the mechanism
+should be considered.
+
+## 3. Should `#[derive_field_bound]` be named just `#[derive_bound]`?
+
+The latter is shorter, but less legible, wherefore we've opted to use
+`#[derive_field_bound]` at the moment.
+
+## 4. Should the attributes not be prefixed with `derive_`?
+
+Prefixing with `derive_` is more legible and reduces the chance of conflict.
+But it is also more verbose, especially when applied on type parameters.
+The current thinking is that readability takes precedence over reducing
+possible verbosity. In any case, prefixing with `derive_` is far less verbose
+than manually implementing the trait.
+
+## 5. Permit `field: Vec<#[derive_field_bound] Arc<T>>`?
+
+If so, `#[derive_bound]` is a more correct name. However, the current thinking
+is that this requires parsing changes while also looking weird. This may
+be a step too far - in such cases, manual `impl`s are probably better.
+For these reasons, the RFC does not propose this mechanism currently.
+
+## 6. Permit `#[derive_bound(<List of traits>, T: <Bound>)]`?
+
+[serde_bound_desc]: https://serde.rs/container-attrs.html#serdebound--t-mytrait
+
+Last but not least, the crate `serde` allows the attribute
+`#[serde(bound = "T: MyBound")]` which replaces the `where`
+clause of the `impl` generated by `serde`. This attribute
+is [described][serde_bound_desc] as follows:
+
+> Where-clause for the `Serialize` and `Deserialize` impls.
+> This replaces any trait bounds inferred by Serde.
+
+We could standardize this concept in the form of an attribute
+`#[derive_bound(..)]` put on types with a syntax permitting:
+
++ Replace bounds on impl of `Clone` and `PartialEq` with `T: Sync`
+
+```rust
+#[derive_bound(Clone, PartialEq, T: Sync)]
```

Comment:
Attributes currently doesn't support this syntax (`T: Sync`).

Reply:
Ah yes, this is what was meant by:
> Changing this would require a language change. 

I'll clarify =)

Reply:
This will be quite a large language change 🙂 And I wonder if this will leads to things like

```
#[derive_bound(T: PartialEq<#[naked] fn(i32)>)]
```

Reply:
On the other hand I think steps to at least provide a sensible subset of things such as bounds, expressions, types in the attribute grammar is good as people already do this kinds of things in serde, derivative, (my derive crate for proptest also does this, but it is still WIP)... It is just done in string quotes which is somewhat hacky :(

That is, afaik, people can already write:

```rust
#[serde(bound = "T: PartialEq<#[naked] fn(i32)>"]
```

I think we should provide custom derive macro authors and users to do these things in a less hacky way.

Reply:
Forgot to say that I clarified this =)

Reply:
@clarcharr I don't think so, but I'm not sure - I'd love to be proven wrong =)

Reply:
Attributes do allow arbitrary token trees between their brackets now.